### PR TITLE
BIP120, Proof of Payment

### PR DESCRIPTION
**Pull request by kallerosenbaum**

### CODE REVIEW

**Comment by kallerosenbaum** [comment]
I have not been able to test NFC, but I have added an intentfilter for btcpop: URIs.

**Comment by schildbach** [comment]
Thanks for the contribution! I'll review your code soon, but before that I'd like to try it. Do you have a test system running somewhere, preferably for testnet?

**Comment by schildbach** [comment]
Two very broad first review comments:
- Can you add the license header to your new files?
- Could you adhere to the code style of the project? Most notably curly braces at the beginning of the line, and usage of the final keyword for "variables" that never change. It's not strictly necessary, I can also reformat your contribution prior to merging if you prefer.

**Comment by kallerosenbaum** [comment]
Ok. Not today, though, but hopefully this weekend.

Test system is available at http://www.rosenbaum.se:8080

There are links to both testnet and prod versions on that page.

/Kalle

2015-11-06 15:15 GMT+01:00 Andreas Schildbach <notifications@github.com>:

**Comment by kallerosenbaum** [comment]
Another thing, I will also update the jpop library to 0.4. It's currently
at 0.4-SNAPSHOT. I had forgotten about it. That will also be done this
weekend.

/Kalle

2015-11-06 15:20 GMT+01:00 Kalle Rosenbaum <kalle@rosenbaum.se>:

**Comment by kallerosenbaum** [comment]
Regarding code style, I imported the eclipse-code-format.xml file into intellij idea using a plugin, Eclipse Code Formatter. I hope it did the job OK for me. At least it seems so.

**Comment by kallerosenbaum** [comment]
NFC is still not tested, I think I need help there. All ask in the forum.

**Comment by schildbach** [comment]
I can test NFC, I have several tags around.

File: wallet/res/layout/pop_content.xml [hunk 1]
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ScrollView xmlns:android="http://schemas.android.com/apk/res/android"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:background="@color/bg_form"
+            android:fillViewport="true"
+            android:scrollbars="none">
+
+    <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (schildbach) [inline wallet/res/layout/pop_content.xml]: No need to repeat the namespace here.

File: wallet/res/layout/pop_content.xml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content"
+                android:drawableLeft="@drawable/ic_warning_grey600_24dp"
+                android:layout_marginTop="8dp"
+                android:layout_marginLeft="@dimen/list_entry_padding_horizontal_lax"
+                android:layout_marginRight="@dimen/list_entry_padding_horizontal_lax"
+                android:text="@string/pop_insecure_warning"/>
+
+        <LinearLayout
+                android:id="@+id/send_coins_private_key_password_group"
```
Comment (schildbach) [inline wallet/res/layout/pop_content.xml]: Please don't re-use a layout ID from another layout. (several occurences)

File: wallet/res/values/strings.xml [hunk 3]
```diff
@@ -313,4 +313,22 @@
...
+	<string name="pop_transaction_information">Prove transaction</string>
+	<string name="pop_transaction_date">Date</string>
+	<string name="pop_transaction_amount">Amount</string>
+	<string name="pop_transaction_memo">Memo</string>
+	<string name="pop_destination">Prove to</string>
+	<string name="pop_insecure_warning">The proof will be sent over an unencrypted connection.</string>
+	<string name="pop_no_matching_transaction">No transaction found matching hints: %s</string>
+	<string name="pop_sent_success">Proof sent successfully</string>
+	<string name="pop_send_invalid_pop">Proof is not valid: %s</string>
+	<string name="pop_send_failed">Failed to send Proof: %s</string>
```
Comment (schildbach) [inline wallet/res/values/strings.xml]: Small 'p'?

File: wallet/res/values/strings.xml [hunk 4]
```diff
@@ -313,4 +313,22 @@
...
+	<string name="pop_input_parser_invalid_btcpop_uri">Invalid Proof of Payment URI: %s</string>
+	<string name="pop_send_pop">Prove</string>
+	<string name="pop_sent">Sent...</string>
```
Comment (schildbach) [inline wallet/res/values/strings.xml]: Can you use the '…' character (AltGr-. on Linux) for ellipsis?

File: wallet/src/de/schildbach/wallet/data/PopIntent.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2011-2015 the original author or authors.
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/data/PopIntent.java]: This copyright is 2015 only, no?

File: wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+				finish();
+				return;
+			}
+		}
+		popRequestURI = popIntent.getPopRequestURI();
+
+		WalletApplication walletApplication = getWalletApplication();
+		Wallet wallet = walletApplication.getWallet();
+
+		List<Transaction> transactions = wallet.getTransactionsByTime();
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java]: I think this section of code could go into a private method (something like "determineMatchingTransaction")

File: wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+			}
+			if (popRequestURI.getLabel() != null && !popRequestURI.getLabel().equals(transaction.getMemo()))
+			{
+				continue;
+			}
+			transactionToProve = transaction;
+			break;
+		}
+
+		if (transactionToProve == null)
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java]: I guess this section is not meant to stay, right? A longToast is still much to short to read all that stuff, and it won't mean much to the user anyway.

File: wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+	@Override
+	public boolean onOptionsItemSelected(final MenuItem item)
+	{
+		switch (item.getItemId())
+		{
+			case android.R.id.home:
+				finish();
+				return true;
+
+			case R.id.send_coins_options_help:
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java]: Again, needs own menu id.

File: wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+		privateKeyPasswordView.setEnabled(newState == State.INPUT);
+	}
+
+	private class Outcome
+	{
+		PopSender popSender;
+		Exception exception;
+	}
+
+	private class SendPopTask extends AsyncTask<PopRequestURI, State, Outcome>
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java]: I generally avoid AsyncTasks in my apps. They're a pain to use (e.g. return values) and produce many spurious errors. Instead, I'm using a background handler, like in the original SendCoinsFragment.

**Comment by schildbach** [comment]
I tested your code and it worked well, at least in the straight forward non-error path. I didn't test:

- NFC – will do that and I don't expect any serious problems.
- Login cases in which I can't provide proof – maybe you could add some of such cases to your test server?

Generally I wonder if it's time to think about some nice abstraction for screens that create and sign transactions. We have SendCoinsFragment, SweepWalletFragment and now PopActivity. Just as a vision: I recently received a Trezor and was thinking of adding support for it as an external signing device. It would be a pity if the effort needs to be repeated for all of those screens. I will think about it a bit, but if you've got ideas let me know.

**Comment by schildbach** [comment]
More comments:
- You might want to add BIP120 to README.specs.
- Activity layout looks a bit rough, but that's ok for a first POC.
- The warning about sending via an insecure link, how important is that, how sensitive is the data? Currently with BIP70 we don't have such a warning.

**Comment by kallerosenbaum** [comment]
I think I've covered all your comments. Great comments by the way, thank you!

I have made some minor GUI improvements, but I still agree with you that it's a bit rough. Suggestions are very appreciated.

Regarding the warning with sending over insecure connection. The pop can be intercepted, but not tampered with. The attacker can take the PoP and use it, but he must either use it pretty much immediately, since the nonce and destination URL will only be valid for a short while (typically 10 minutes or so). I can remove the warning if you think it's appropriate. We still have the padlock on the destination host to indicate security to the user.

**Comment by kallerosenbaum** [comment]
I see we have conflicts now. Do you want me to rebase? Is there a problem with this pull req that keeps it from getting merged (apart from the recent conflict)?

**Comment by schildbach** [comment]
Yes, rebasing is always appreciated.

As long as it's in the POC phase I'd rather not merge it to master. I can merge it to a branch if that's useful to you, and build a version signed with the release key to people can switch back and forth without the need to uninstall.

**Comment by kallerosenbaum** [comment]
I'm not sure what the way forward is here. You say this is a POC, but what does that mean in this context? Is this a phase where we gather feedback from other people regarding the GUI? What am I supposed to do?

If we need feedback from other people, then yes, I'd appreciate a branch where you build and sign so people can test it without uninstalling. Otherwise I don't think we need a branch. Do you? If you're merging to a branch, please wait until I've rebased.

**Comment by kallerosenbaum** [comment]
Rebased!

File: wallet/src/de/schildbach/wallet/data/PopIntent.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+	@Override
+	public void writeToParcel(final Parcel parcel, int i)
+	{
+		parcel.writeString(popRequestURI.toURIString());
+	}
+
+	public static final Parcelable.Creator<PopIntent> CREATOR = new Parcelable.Creator<PopIntent>()
+	{
+		public PopIntent createFromParcel(final Parcel in)
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/data/PopIntent.java]: This needs @Override I think.

File: wallet/src/de/schildbach/wallet/data/PopIntent.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		{
+			return new PopIntent(in);
+		}
+
+		public PopIntent[] newArray(final int size)
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/data/PopIntent.java]: This needs @Override I think.

File: wallet/src/de/schildbach/wallet/ui/pop/SendPopTask.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
+package de.schildbach.wallet.ui.pop;
```
Comment (schildbach) [inline wallet/src/de/schildbach/wallet/ui/pop/SendPopTask.java]: This file is missing the license (and author, if you want).

**Comment by kallerosenbaum** [comment]
Closing this since there is a new pull-req for bip120 (https://github.com/schildbach/bitcoin-wallet/pull/272).

**Comment by benny5181** [comment]
I'm exchange my money for bit coin

**Pull request closed without merging**